deploy/olm-catalog/service-telemetry-operator/metadata/properties.yaml

```diff
@@ -1,3 +1,44 @@
 properties:
   - type: olm.maxOpenShiftVersion
     value: "4.12"
+  - type: olm.constraint
+    value:
+      failureMessage: Require Smart Gateway for Service Telemetry Framework
+      all:
+        constraints:
+          - failureMessage: Package smart-gateway-operator is needed for Service Telemetry Framework
+            package:
+              packageName: smart-gateway-operator
+              versionRange: '>=5.0.0'
+  - type: olm.constraint
+    value:
+      failureMessage: Require data transport for Service Telemetry Framework
+      all:
+        constraints:
+          - failureMessage: Package amq7-interconnect-operator is needed for data transport with STF
+            package:
+              packageName: amq7-interconnect-operator
+              versionRange: '>=1.10.0'
+  - type: olm.constraint
+    value:
+      failureMessage: Require certificate management for Service Telemetry Framework
+      all:
+        constraints:
+          - failureMessage: Package openshift-cert-manager-operator is needed for AMQ Interconnect setup
+            package:
+              packageName: openshift-cert-manager-operator
+              versionRange: '>=1.10.0'
+  - type: olm.constraint
+    value:
+      failureMessage: Require Prometheus backend for data storage of metrics for Service Telemetry Framework
+      any:
+        constraints:
+          - package:
+              packageName: prometheus
+              versionRange: '>= 0.56.0'
+          - package:
+              packageName: rhods-prometheus-operator
+              versionRange: '>=4.10.0'
+          - package:
```

Comment:
I can't promise that, and I was also very surprised :)

Reply:
I don't think we want to add any logic to the operator itself to do the dependency management. From what I read, it's going to try and satisfy from the bottom up for package resolution. Priority for same-named packages is based on the priority order applied to the CatalogSource itself (higher number in the CatalogSource is preferred to satisfy for the package).

The idea here is that OBO will be the preferred installation during Service Telemetry Operator deployment. I believe I saw some code that would check for the availability of the API interface? If not, that's probably what we'd want to add (we have a few of them already in there checking that the CRD is available before attempting to create objects in OCP against that API).

So I would expect that if you set observabilityStrategy: use_redhat, but OBO is not installed, then STO should skip the creation of the Prometheus objects against the OBO CRDs since it's not available, and you'd just end up without a Prometheus deployment.

I think it might be worth going through our Operator again and adding some status updates to the ServiceTelemetry object to provide some better debugging values in the future. We could add a status that says something like OBO CRD is not available when the strategy is set to use_redhat, to make it more clear why someone might not get a Prometheus deployment.

Reply:
> Maybe we could add some logic to evaluate the observability_strategy in use to pick the specific dependency?

These deps are checked and satisfied before an STF object with an observability_strategy ever exists. The observability_strategy allows us to pick between the two different APIs, but I agree that we don't want to actually install or remove other operators from our code - that's OLM's job.

Reply:
What I was thinking here was doing an API check though like here (https://github.com/infrawatch/service-telemetry-operator/blob/master/roles/servicetelemetry/tasks/main.yml#L10-L23) where if `use_redhat` is enabled, but the OBO CRD isn't loaded on the system, then we can skip over the creation of the OBO objects.